pennylane/transforms/optimization.py

```diff
@@ -0,0 +1,439 @@
+
+    Consider the following quantum function.
+
+    .. code-block:: python
+
+        def qfunc(r1, r2):
+            qml.Hadamard(wires=0)
+            qml.Rot(*r1, wires=0)
+            qml.Rot(*r2, wires=0)
+            qml.RZ(r1[0], wires=0)
+            qml.RZ(r2[0], wires=0)
+            return qml.expval(qml.PauliX(0))
+
+    The circuit before optimization:
+
+    >>> dev = qml.device('default.qubit', wires=1)
+    >>> qnode = qml.QNode(qfunc, dev)
+    >>> print(qml.draw(qnode)([0.1, 0.2, 0.3], [0.4, 0.5, 0.6]))
+    0: ──H──Rot(0.1, 0.2, 0.3)──Rot(0.4, 0.5, 0.6)──RZ(0.1)──RZ(0.4)──┤ ⟨X⟩
+
+    Full single-qubit gate fusion allows us to collapse this entire sequence into a
+    single ``qml.Rot`` rotation gate.
+
+    >>> optimized_qfunc = qml.compile(pipeline=[single_qubit_fusion], num_passes=3)(qfunc)
+    >>> optimized_qnode = qml.QNode(optimized_qfunc, dev)
+    >>> print(qml.draw(optimized_qnode)([0.1, 0.2, 0.3], [0.4, 0.5, 0.6]))
+    0: ──Rot(3.57, 2.09, 2.05)──┤ ⟨X⟩
+
+    """
+    # Make a working copy of the list to traverse
+    list_copy = tape.operations.copy()
+
+    while len(list_copy) > 0:
+        current_gate = list_copy[0]
+
+        # Normally queue any multi-qubit gates
+        if current_gate.num_wires > 1:
+            current_gate.queue()
+            list_copy.pop(0)
+            continue
+
+        # Find the next gate that acts on the same wires
+        next_gate_idx = _find_next_gate(current_gate.wires, list_copy[1:])
+
+        # If no such gate is found (either there simply is none, or there are other gates
+        # "in the way", queue the operation and move on
+        if next_gate_idx is None:
+            current_gate.queue()
+            list_copy.pop(0)
+            continue
+
+        # Set up a cumulative Rot starting from the angles of the initial gate
+        cumulative_angles = current_gate.as_rot_angles()
+
+        # Loop as long as a valid next gate exists
+        while next_gate_idx is not None:
+            # Get the next gate
+            next_gate = list_copy[next_gate_idx + 1]
+
+            # If next gate is on the same qubit, we can fuse them
+            if current_gate.wires == next_gate.wires:
+                list_copy.pop(next_gate_idx + 1)
+
+                # Merge the angles
+                next_gate_angles = next_gate.as_rot_angles()
+                cumulative_angles = fuse_rot(cumulative_angles, next_gate_angles)
+            else:
+                break
+
+            next_gate_idx = _find_next_gate(current_gate.wires, list_copy[1:])
+
+        # If the cumulative angle is not close to 0, apply the gate
+        if not allclose(np.array(cumulative_angles), np.zeros(3)):
+            Rot(*cumulative_angles, wires=current_gate.wires).queue()
+
+        # Remove the starting gate from the list
+        list_copy.pop(0)
+
+    # Queue the measurements normally
+    for m in tape.measurements:
+        m.queue()
+
+
+@qfunc_transform
+def diag_behind_controls(tape):
```

Comment:
I think this function needs a better name; it may also be better to rewrite it "backwards". Right now, the pass stops at each CNOT gate, and looks ahead to see if any upcoming single-qubit gates can be pushed behind. But we could alternatively stop at the diagonal single-qubit gates, and look behind to see if they can be pushed through anything.